akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -210,6 +210,166 @@ final class PrefixAndTail[T](n: Int) extends GraphStage[FlowShape[T, (immutable.
+
+    override def preStart(): Unit = timeout = ActorMaterializer.downcast(interpreter.materializer).settings.subscriptionTimeoutSettings.timeout
+
+    override def onPull(): Unit = {
+      substreamWaitingToBePushed match {
+        case Some(substreamSource) ⇒
+          push(out, Source.fromGraph(substreamSource.source))
+          scheduleOnce(substreamSource.key, timeout)
+          substreamWaitingToBePushed = None
+        case None ⇒
+          if (hasNextElement) {
+            activeSubstreamsMap(nextElement._1).push(nextElement._2)
+            nextElement = null
+          } else tryPull(in)
+      }
+    }
+
+    override def onUpstreamFailure(ex: Throwable): Unit = fail(ex)
+
+    override def onDownstreamFinish(): Unit =
+      if (activeSubstreamsMap.isEmpty) completeStage() else setKeepGoing(true)
+
+    override def onPush(): Unit = {
+      val elem = grab(in)
+      val key: K = try {
+        val k = f(elem)
+        require(k != null, "Key cannot be null")
+        k
+      } catch {
+        case NonFatal(ex) ⇒
+          decider(ex) match {
+            case Supervision.Stop                         ⇒ fail(ex)
+            case Supervision.Resume | Supervision.Restart ⇒ if (!hasBeenPulled(in)) pull(in)
+          }
+          null.asInstanceOf[K]
+      }
+
+      if (key != null) {
+        activeSubstreamsMap.get(key) match {
+          case Some(substreamSource) ⇒
+            if (substreamSource.isAvailable) substreamSource.push(elem)
+            else nextElement = key -> elem
+          case None ⇒
+            if (activeSubstreamsMap.size == maxSubstreams)
+              fail(new IllegalStateException(s"Cannot open substream for key '$key': too many substreams open"))
+            else if (closedSubstreams(key) && !hasBeenPulled(in)) pull(in)
+            else runSubstream(key, elem)
+        }
+      }
+    }
+
+    override def onUpstreamFinish(): Unit = {
+      if (!completeWholeStage()) setKeepGoing(true)
+    }
+
+    private def runSubstream(key: K, value: T): Unit = {
+      val substreamSource = new SubstreamSource("GroupBySource " + nextId, key, value)
+
+      activeSubstreamsMap += key -> substreamSource
+
+      if (isAvailable(out)) {
+        push(out, Source.fromGraph(substreamSource.source))
+        scheduleOnce(key, timeout)
+        substreamWaitingToBePushed = None
+      } else {
+        setKeepGoing(true)
+        substreamWaitingToBePushed = Some(substreamSource)
+      }
+    }
+
+    override protected def onTimer(timerKey: Any): Unit = {
+      activeSubstreamsMap.get(timerKey).foreach(substreamSource ⇒ {
+        substreamSource.timeout(timeout)
+        closedSubstreams += timerKey
+        activeSubstreamsMap -= timerKey
+      })
+      if (isClosed(in)) completeWholeStage()
+    }
+
+    setHandlers(in, out, this)
+
+    private class SubstreamSource(name: String, val key: K, val firstElement: T) extends SubSourceOutlet[T](name) {
```

Comment:
This will keep the first element in memory as long as the stream lives, should probably be `null` ed after consumed, then `firstPush` can be a `def firstPush = firstElement != null` instead.